platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+      pingController?.shutdown()
+
+    where:
+      // return_or_throw -> null to return null, exception to be thrown, SolrClient to return a client
+      created_after                                                    || return_or_throw
+      '1 attempt'                                                      || SolrClient
+      '2 attempts where the first one returned null'                   || null
+      '2 attempts where the first one failed with SolrServerException' || SolrServerException
+      '2 attempts where the first one failed with SolrException'       || new SolrException(
+          ErrorCode.UNKNOWN, 'testing')
+      '2 attempts where the first one failed with IOException'         || IOException
+      '2 attempts where the first one failed with RutimeException'     || RuntimeException
+  }
+
+  @Unroll
+  // the following needs to be changed as it should be closing the whole thing and not re-execute the creation
+  def 'test constructor when client is created after 2 attempts where the first one is interrupted with #exception.simpleName and becomes available'() {
+    given: "a client and creator to verify that failsafe actually delegates to our creator and client"
+      def client = Mock(SolrClient) {
+        1 * ping() >> pingOkResponse
+      }
+      def creator = Mock(Creator) {
+        1 * create() >> client
+      }
+
+    and: "controllers that will actually proceed through Solr"
+      def createController = new FailsafeController('SolrClient Creation')
+          .onNextExecution(doThrow(exception))
+          .and().onNextExecution(doProceed())
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution(doProceed())
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the reported core should correspond to the requested one"
+      adapter.core == CORE
+
+    and:
+      "our creator should be registered with the adapter"
+      adapter.creator.is(creator)
+
+    when: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.isAvailable()
+
+    and: "the underlying client should be the one we created"
+      adapter.proxiedClient.is(client)
+
+    and: "the underlying client should never be closed"
+      0 * client.close()
+
+    and: "there should be no registered initializers or listeners"
+      !adapter.hasInitializers() && !adapter.hasListeners()
+
+    and: "verify failsafe controllers"
+      createController.verify()
+      pingController.verify()
+
+    cleanup:
+      createController?.shutdown()
+      pingController?.shutdown()
+
+    where:
+      exception << [InterruptedException, InterruptedIOException]
+  }
+
+  @Unroll
+  def 'test client #test_client_how where an API call also fails to connect after the initial failure'() {
```

Comment:
❓Can we improve the description to make it clear that these tests are verifying that the adaptor recovers from an initial failure?

Reply:
I can. Remember that the whole design is for the adapter to always recover so I omitted that from the test cases and only talked about the error conditions being tested instead

Reply:
Since the tests are fairly involved and the descriptions can be as long as we need them to be, I think we should make them as clear and explicit as possible. I'm sure it'd be _very_ helpful for someone looking at these tests for the first time.

Reply:
done